### PR TITLE
Create an UTF8-Checker

### PR DESCRIPTION
**Pull request by felix-re**
Create an utf8-checker for text and byte sequences and unit-tests for it. The websocket-peer can use it to validate the incoming text messages and frames. Add it qbs and cmake build files.

### CODE REVIEW

**Comment by felix-re** [comment]
I'm fixing the build problems

**Comment by felix-re** [comment]
All build problems solved. Ready to review.

File: src/tests/utf8_checker_test.cpp [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+#define BOOST_TEST_MODULE utf8_checker_tests
+
+#include <boost/test/unit_test.hpp>
+
+#include "utf8_checker.h"
+
+static const int invalid_message_size = 9;
+static const uint8_t invalid_message[invalid_message_size][4] = {{0xC0,0x00,0x00,0x00},	//invalid start
+																 {0xF6,0x00,0x00,0x00},	//invalid start
+																 {0x80,0x00,0x00,0x00},	//invalid start
```
Comment (gatzka) [inline src/tests/utf8_checker_test.cpp]: What an awful bunch of empty lines.
Please reduce that.

File: src/tests/utf8_checker_test.cpp [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+ * @brief tests the utf8 checker with a fragmented valid message
+ *
+ * The message consists of two 4 byte utf8 characters. The message
+ * is fragmented after each byte.
+ */
+BOOST_AUTO_TEST_CASE(test_valid_message_fragmented_between_letters_long)
+{
+	F f;
+	int ret;
+	for (uint i = 0; i < sizeof(valid_message_long) - 1; i++) {
```
Comment (gatzka) [inline src/tests/utf8_checker_test.cpp]: Why uint and not unsigned int?

File: src/tests/utf8_checker_test.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+	BOOST_CHECK_MESSAGE(ret == true, "Message should be valid!");
+}
+
+/**
+ * @brief tests the utf8 checker with a invalid message
+ */
+BOOST_AUTO_TEST_CASE(test_invalid_message)
+{
+	int ret;
+	for (int i = 0; i < invalid_message_size; i++) {
```
Comment (gatzka) [inline src/tests/utf8_checker_test.cpp]: Please use an unsigned int also.

File: src/tests/utf8_checker_test.cpp [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+ * @brief tests the utf8 checker with a fragmented invalid message
+ *
+ * The message is fragmented after each byte. Hence some utf8 characters
+ * are fragmented, too.
+ */
+BOOST_AUTO_TEST_CASE(test_invalid_message_fragmented_between_letters)
+{
+	int ret;
+	bool invalid = false;
+	for (int i = 0; i < invalid_message_size; i++) {
```
Comment (gatzka) [inline src/tests/utf8_checker_test.cpp]: unsigned int.

File: src/utf8_checker.c [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "utf8_checker.h"
```
Comment (gatzka) [inline src/utf8_checker.c]: include <stdbool.h>, >stdint> is missing.

File: src/utf8_checker.c [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+#define UC_FINISH 0xFF
```
Comment (gatzka) [inline src/utf8_checker.c]: Possible to use an const uint8_t?

File: src/utf8_checker.c [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+static bool is_byte_valid(struct cjet_utf8_checker *c, uint8_t byte)
+{
+	bool ret = true;
+	int finished = 0;
```
Comment (gatzka) [inline src/utf8_checker.c]: Why not using a bool?

File: src/utf8_checker.c [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+	} else {
+		c->next_byte++;
+	}
+	return ret;
+}
+
+bool cjet_is_text_valid(struct cjet_utf8_checker *c, const char *text, size_t length, bool is_complete)
+{
+	bool ret = true;
+	for (unsigned int i = 0; i < length; i++) {
```
Comment (gatzka) [inline src/utf8_checker.c]: Please use an size_t, to have the same type when comparing to length.

**Review by gatzka — CHANGES_REQUESTED** [review]

File: src/tests/utf8_checker_test.cpp (hunk 1 above)
Comment (felix-re) [reply src/tests/utf8_checker_test.cpp]: replaced tabs with spaces

File: src/tests/utf8_checker_test.cpp (hunk 2 above)
Comment (felix-re) [reply src/tests/utf8_checker_test.cpp]: Didn't know, that it's just a typedef.

File: src/tests/utf8_checker_test.cpp (hunk 3 above)
Comment (felix-re) [reply src/tests/utf8_checker_test.cpp]: done

File: src/tests/utf8_checker_test.cpp (hunk 4 above)
Comment (felix-re) [reply src/tests/utf8_checker_test.cpp]: done

File: src/utf8_checker.c (hunk 5 above)
Comment (felix-re) [reply src/utf8_checker.c]: Included, but isn't it sufficient to include it in the header?

File: src/utf8_checker.c (hunk 6 above)
Comment (felix-re) [reply src/utf8_checker.c]: yes

File: src/utf8_checker.c (hunk 7 above)
Comment (felix-re) [reply src/utf8_checker.c]: Overlooked, when included stdbool.h

File: src/utf8_checker.c (hunk 8 above)
Comment (felix-re) [reply src/utf8_checker.c]: done

**Review by gatzka — APPROVED** [review]

**Pull request merged**